File: 백준/Gold/1600. 말이 되고픈 원숭이/말이 되고픈 원숭이.py
from collections import deque

K = int(input())
M, N = map(int, input().split())
arr = [list(map(int, input().split())) for _ in range(N)]
visited = [[[-1]*(K+1) for _ in range(M)] for _ in range(N)]
dx, dy = [0, 1, 0, -1], [1, 0, -1, 0]
hdx, hdy = [-2, -1, 1, 2, 2, 1, -1, -2], [1, 2, 2, 1, -1, -2, -2, -1]

def bfs():
    q = deque([(0, 0, 0)])
    visited[0][0][0] = 0

    while q:
        x, y, use = q.popleft()

        if use < K:
            for i in range(8):
                nx, ny = x+hdx[i], y+hdy[i]
                if 0<=nx<N and 0<=ny<M and arr[nx][ny]!=1 and visited[nx][ny][use+1]==-1:
                    q.append((nx, ny, use+1))
                    visited[nx][ny][use+1] = visited[x][y][use] + 1
        for i in range(4):
            nx, ny = x+dx[i], y+dy[i]
            if 0<=nx<N and 0<=ny<M and arr[nx][ny]!=1 and visited[nx][ny][use]==-1:
                q.append((nx, ny, use))
                visited[nx][ny][use] = visited[x][y][use] + 1
    result = [visited[N-1][M-1][i] for i in range(K+1) if visited[N-1][M-1][i] != -1]
    return min(result) if result else -1
print(bfs())

File: 백준/Gold/1600. 말이 되고픈 원숭이/말이 되고픈 원숭이.cc
#include <iostream>
#include <queue>
#include <vector>
#include <tuple>

using namespace std;

int K, N, M;
int dx[4] = {0, 1, 0, -1};
int dy[4] = {1, 0, -1, 0};
int hx[8] = {-2, -1, 1, 2, 2, 1, -1, -2};
int hy[8] = {1, 2, 2, 1, -1, -2, -2, -1};

vector<vector<int>> arr;
vector<vector<vector<int>>> visited;

int bfs(){
    queue<tuple<int, int, int>> q;
    q.push({0, 0, K});
    visited[0][0][K] = 0;

    while(!q.empty()){
        auto[x, y, rest] = q.front();
        q.pop();
        if(x==N-1 && y==M-1){
            return visited[x][y][rest];
        }

        if(rest>0){
            for(int i=0; i<8; i++){
                int nx = x+hx[i];
                int ny = y+hy[i];
                if(0<=nx && nx<N && 0<=ny && ny<M && visited[nx][ny][rest-1]==-1 && arr[nx][ny]==0){
                    q.push({nx, ny, rest-1});
                    visited[nx][ny][rest-1] = visited[x][y][rest] + 1;
                }
            }
        }
        for(int i=0; i<4; i++){
            int nx = x+dx[i];
            int ny = y+dy[i];
            if(0<=nx && nx<N && 0<=ny && ny<M && visited[nx][ny][rest]==-1 && arr[nx][ny]==0){
                q.push({nx, ny, rest});
                visited[nx][ny][rest] = visited[x][y][rest] + 1;
            }
        }
    }
    return -1;

}

int main(void){
    cin >> K >> M >> N;

    arr.resize(N, vector<int>(M, 0));
    visited.resize(N, vector<vector<int>>(M, vector<int>(K+1, -1)));

    for(int i=0; i<N; i++){
        for(int j=0; j<M; j++){
            cin >> arr[i][j];
        }
    }
    
    int result = bfs();

    cout << result << '\n';

    return 0;
}